apis/python/src/tiledbsoma/soma_array.cc

```diff
@@ -0,0 +1,506 @@
+}
+
+void load_soma_array(py::module &m) {
+    py::class_<SOMAArray>(m, "SOMAArray")
+        .def(
+            py::init(
+                [](std::string_view uri,
+                   std::string_view name,
+                   std::optional<std::vector<std::string>> column_names_in,
+                   std::string_view batch_size,
+                   ResultOrder result_order,
+                   std::map<std::string, std::string> platform_config,
+                   std::optional<std::pair<uint64_t, uint64_t>> timestamp) {
+                    // Handle optional args
+                    std::vector<std::string> column_names;
+                    if (column_names_in) {
+                        column_names = *column_names_in;
+                    }
+
+                    return SOMAArray::open(
+                        OpenMode::read,
+                        uri,
+                        name,
+                        platform_config,
+                        column_names,
+                        batch_size,
+                        result_order,
+                        timestamp);
+                }),
+            "uri"_a,
+            py::kw_only(),
+            "name"_a = "unnamed",
+            "column_names"_a = py::none(),
+            "batch_size"_a = "auto",
+            "result_order"_a = ResultOrder::automatic,
+            "platform_config"_a = py::dict(),
+            "timestamp"_a = py::none())
+
+        .def(
+            "set_condition", 
+            [](SOMAArray& reader, 
+               py::object py_query_condition,
+               py::object py_schema){
+                   auto column_names = reader.column_names();
+                   // Handle query condition based on
+                   // TileDB-Py::PyQuery::set_attr_cond()
+                   QueryCondition* qc = nullptr;
+                   if (!py_query_condition.is(py::none())) {
+                       py::object init_pyqc = py_query_condition.attr(
+                           "init_query_condition");   
+                       try {
+                           // Column names will be updated with columns present
+                           // in the query condition
+                           auto new_column_names =
+                               init_pyqc(py_schema, column_names)
+                                   .cast<std::vector<std::string>>();   
+                           // Update the column_names list if it was not empty,
+                           // otherwise continue selecting all columns with an
+                           // empty column_names list
+                           if (!column_names.empty()) {
+                               column_names = new_column_names;
+                           }
+                       } catch (const std::exception& e) {
+                           TPY_ERROR_LOC(e.what());
+                       }   
+                       qc = py_query_condition.attr("c_obj")
```

Comment:
What is the lifetime here?

Reply:
For the lifetime of the `ManagedQuery` / TileDB `Query` associated with the `SOMAArray`. We set the query condition with `SOMAArray::set_condition` -> `ManagedQuery::set_condition` -> `tiledb::Query::set_condition`. Everytime we run a new query, we construct a new `ManagedQuery` object which in turn constructs a new `Query` object.